core/clustersmngr/factory.go

```diff
@@ -0,0 +1,221 @@
+	clustersFetcher ClusterFetcher
+
+	clustersMutex sync.Mutex
+	clusters      []Cluster
+
+	clustersNsMutex    sync.RWMutex
+	clustersNamespaces map[string][]v1.Namespace
+
+	usersNamespaces *UsersNamespaces
+
+	nsChecker nsaccess.Checker
+
+	logger logr.Logger
+}
+
+func NewClientFactory(hubClient client.Client, fetcher ClusterFetcher, nsChecker nsaccess.Checker, logger logr.Logger) *ClientsFactory {
+	return &ClientsFactory{
+		hubClient:          hubClient,
+		clustersFetcher:    fetcher,
+		nsChecker:          nsChecker,
+		clusters:           []Cluster{},
+		clustersNamespaces: map[string][]v1.Namespace{},
+		usersNamespaces:    &UsersNamespaces{Cache: ttlcache.New(24 * time.Hour)},
+		logger:             logger,
+	}
+}
+
+func (cf *ClientsFactory) Start(ctx context.Context) error {
+	go cf.watchClusters(ctx)
+	go cf.watchNamespaces(ctx)
+
+	return nil
+}
+
+func (cf *ClientsFactory) watchClusters(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(30*time.Second, func() (bool, error) {
+		clusters, err := cf.clustersFetcher.Fetch(ctx)
+		if err != nil {
+			return false, fmt.Errorf("failed to fetch clusters: %w", err)
+		}
+
+		cf.clustersMutex.Lock()
+		cf.clusters = clusters
+		cf.clustersMutex.Unlock()
+
+		return true, nil
+	}); err != nil {
+		// TODO: log error
+		return
+	}
+}
+
+func (cf *ClientsFactory) watchNamespaces(ctx context.Context) {
+	if err := wait.PollImmediateInfinite(30*time.Second, func() (bool, error) {
+		clients, err := clientsForClusters(cf.clusters)
+		if err != nil {
+			cf.logger.Error(err, "failed to create clients for", "clusters", cf.clusters)
+			return false, err
+		}
+
+		wg := sync.WaitGroup{}
+
+		for clusterName, c := range clients {
+			wg.Add(1)
+
+			go func(clusterName string, c client.Client) {
+				defer wg.Done()
+
+				nsList := &v1.NamespaceList{}
+
+				if err := c.List(ctx, nsList); err != nil {
+					cf.logger.Error(err, "failed listing namespaces", "cluster", clusterName)
+				}
+
+				cf.clustersNsMutex.Lock()
+				cf.clustersNamespaces[clusterName] = nsList.Items
+				cf.clustersNsMutex.Unlock()
+			}(clusterName, c)
+		}
+
+		wg.Wait()
+
+		return false, nil
+	}); err != nil {
+		cf.logger.Error(err, "failed pooling namespaces")
+	}
+}
+
+func (cf *ClientsFactory) GetUserClient(ctx context.Context, user *auth.UserPrincipal) (Client, error) {
+	pool := NewClustersClientsPool()
+
+	for _, cluster := range cf.clusters {
+		if err := pool.Add(ClientConfigWithUser(user), cluster); err != nil {
+			return nil, fmt.Errorf("failed adding cluster client to pool: %w", err)
+		}
+	}
+
+	return NewClient(pool, cf.userNsList(ctx, user)), nil
+}
+
+func restConfigFromCluster(cluster Cluster) *rest.Config {
+	return &rest.Config{
+		Host:            cluster.Server,
+		BearerToken:     cluster.BearerToken,
+		TLSClientConfig: cluster.TLSConfig,
+	}
+}
+
+func (cf *ClientsFactory) userNsList(ctx context.Context, user *auth.UserPrincipal) map[string][]v1.Namespace {
```

Comment:
How can an endpoint fetch available namespaces? With `cacheContainer` is was exposed not only to access them when making queries, but the `ListNamespaces` and the `GetFluxNamespace` uses the raw list of namespaces.

Reply:
Regarding `ListNamespaces` and `GetFluxNamespace` those are unused and probably introduced early on when we needed something in place to keep going, but they should not be necessary and can be removed.

Reply:
I think we need `GetFluxNamespace` to get the `Flux Runtime` page to work. If we aren't using that now, we should be.